ash/src/lib.rs

```diff
@@ -52,6 +52,32 @@ impl<'r, T> RawPtr<T> for Option<&'r T> {
     }
 }
 
+/// Include correctly aligned and typed precompiled SPIR-V
+///
+/// Does not account for endianness mismatches between the SPIR-V file and the target. See
+/// `util::read_spv` for a more general solution.
+#[macro_export]
+macro_rules! include_spv {
+    ($path:expr) => {
+        &$crate::util::Align4(*include_bytes!($path)) as &$crate::Spirv
```

Comment:
It is a bit unclear to me why this is useful. I don't think this is what the user would expect. 

It is aligned correctly if the len of the byte array is `[u8; LEN], LEN % 4 == 0` right? And this invariant can't break because it would not be legal spir-v otherwise. Loading anything that violates this is not correct and we should probably panic instead.

Reply:
> It is a bit unclear to me why this is useful. I don't think this is what the user would expect.

What specifically are you referring to?

> It is aligned correctly if the len of the byte array is [u8; LEN], LEN % 4 == 0 right?

No, alignment is a property of where it's stored, not its length.

> we should probably panic

That would preclude use in a const context, which is a goal.

Reply:
I am confused. Why do you think it has to be aligned? The vulkan spec only states that the length has to be a multiple of 4 (because 1 spirv op is u32).

Reply:
To begin with, `vk::ShaderModuleCreateInfo` takes a `*const u32`, and the builder takes a `&[u32]`, so at a baseline this is more ergonomic. Moreover, [the spec](https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkShaderModuleCreateInfo.html) says:

>  pCode must be a valid pointer to an array of codeSize / 4 uint32_t values

An unaligned `&[u8]` violates that. Under [the C11 memory model](http://port70.net/~nsz/c/c11/n1570.html#6.3.2.3p7), even *constructing* a `*const u32` from unaligned memory is UB.

We've discussed this previously in https://github.com/MaikKlein/ash/pull/117 and https://github.com/MaikKlein/ash/pull/170.

Reply:
You are totally right.

Reply:
I'll look at it more closely tomorrow. But it seems to me that this copies the spv to the stack right?

Reply:
In principle it could; you can put it in a `const` or even a `static` if you're worried that the blindingly obvious optimization might not happen and your shaders are extremely large.

Reply:
Actually, given that it has `'static` lifetime, that might not be a risk at all.

Reply:
I believe you're right, the spv will only make it to the stack if it is copied.

Reply:
It looks like dereffing `include_bytes` does *technically* count as a copy when not in const context, so probably best to encourage people to stick it in a `const`, though I imagine it'll optimize down the same either way.